sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -0,0 +1,242 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.consumer;
+
+import com.google.common.collect.ImmutableMap;
+import io.crate.analyze.MultiSourceSelect;
+import io.crate.analyze.QuerySpec;
+import io.crate.analyze.WhereClause;
+import io.crate.analyze.relations.JoinPair;
+import io.crate.analyze.relations.QueriedRelation;
+import io.crate.analyze.symbol.FuncSymbols;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.RefReplacer;
+import io.crate.analyze.symbol.SelectSymbol;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.SymbolVisitor;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.table.Operation;
+import io.crate.operation.operator.Operator;
+import io.crate.operation.operator.OrOperator;
+import io.crate.operation.operator.any.AnyOperator;
+import io.crate.operation.predicate.NotPredicate;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.sql.tree.QualifiedName;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static io.crate.analyze.expressions.ExpressionAnalyzer.castIfNeededOrFail;
+import static io.crate.operation.operator.Operators.LOGICAL_OPERATORS;
+
+public class SemiJoins {
+
+    /**
+     * Try to rewrite a QueriedRelation into a SemiJoin:
+     *
+     * <pre>
+     *
+     *     select x from t1 where x in (select id from t2)
+     *               |
+     *               v
+     *     select t1.x from t1 SEMI JOIN (select id from t2) t2 on t1.x = t2.id
+     * </pre>
+     *
+     * (Note that it's not possible to write a SemiJoin directly using SQL. But the example above is semantically close.)
+     *
+     * This rewrite isn't always possible - certain conditions need to be met for the semi-join to have the same
+     * semantics as the subquery.
+     *
+     * @return the rewritten relation or null if a rewrite wasn't possible.
+     */
+    @Nullable
+    public static QueriedRelation tryRewrite(QueriedRelation rel) {
+        WhereClause where = rel.querySpec().where();
+        if (!where.hasQuery()) {
+            return null;
+        }
+        List<Function> rewriteCandidates = gatherRewriteCandidates(where.query());
+        if (rewriteCandidates.isEmpty()) {
+            return null;
+        }
+        if (rewriteCandidates.size() > 1) {
+            return null;  // TODO: support this as well
+        }
+        Function rewriteCandidate = rewriteCandidates.get(0);
+        SelectSymbol selectSymbol = getSubqueryOrNull(rewriteCandidate.arguments().get(1));
+        assert selectSymbol != null : "rewriteCandidate must contain a selectSymbol";
+
+        removeRewriteCandidatesFromWhere(rel, rewriteCandidate);
```

Comment:
Fully agree - I think we should make the relations/querySpec immutable as well - and also consider merging querySpec into the relations to be able to work more with polymorphism instead of visitors.

But that's a bigger change. And for now it's not simple to create a new QueriedRelation since it is an interface.